### PR TITLE
Add spiral-matrix

### PR DESCRIPTION
**Pull request by ryanplusplus**
Related to https://github.com/exercism/c/issues/940

### CODE REVIEW

File: exercises/practice/spiral-matrix/test_spiral_matrix.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                           const spiral_matrix_t *actual)
+{
+   if (expected_size != actual->size)
+      return false;
+
+   for (int i = 0; i < expected_size; i++) {
+      for (int j = 0; j < expected_size; j++) {
+         if (expected_matrix[i][j] != actual->matrix[i][j]) {
+            printf("expected[%d][%d] = %d\n", i, j, expected_matrix[i][j]);
+            printf("actual[%d][%d] = %d\n", i, j, actual->matrix[i][j]);
```
Comment (ryanplusplus) [inline exercises/practice/spiral-matrix/test_spiral_matrix.c:26]: Because the check is pretty opaque (it tells you whether it matches, but that's it), I added in these prints to show where the first difference is when there's a failure.

File: exercises/practice/spiral-matrix/test_spiral_matrix.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+static void test_spiral_of_size_5(void)
+{
+   TEST_IGNORE();
+
+   int expected_matrix[5][5] = {
+      { 1, 2, 3, 4, 5 },       //
+      { 16, 17, 18, 19, 6 },   //
+      { 15, 24, 25, 20, 7 },   //
+      { 14, 23, 22, 21, 8 },   //
```
Comment (ryanplusplus) [inline exercises/practice/spiral-matrix/test_spiral_matrix.c:109]: These `//`s prevent the formatter from wrapping these with more than one row on a line. I only need these when there are 5 rows.

File: exercises/practice/spiral-matrix/.meta/example.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   for (int i = 0; i < size; i++) {
+      spiral_matrix->matrix[i] = calloc(size, sizeof(int));
+   }
+
+   int x = 0;
+   int y = 0;
+   int dx = 1;
+   int dy = 0;
+
+   for (int i = 1; i <= size * size; i++) {
```
Comment (wolf99) [inline exercises/practice/spiral-matrix/.meta/example.c:23]: `size_t i` maybe?

File: exercises/practice/spiral-matrix/.meta/example.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      x += dx;
+      y += dy;
+   }
+
+   return spiral_matrix;
+}
+
+void spiral_matrix_destroy(spiral_matrix_t *spiral_matrix)
+{
+   for (int i = 0; i < spiral_matrix->size; i++) {
```
Comment (wolf99) [inline exercises/practice/spiral-matrix/.meta/example.c:41]: `size_t i` maybe?

File: exercises/practice/spiral-matrix/spiral_matrix.h [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#ifndef SPIRAL_MATRIX_H
+#define SPIRAL_MATRIX_H
+
```
Comment (wolf99) [inline exercises/practice/spiral-matrix/spiral_matrix.h:3]: Given the recent feedbac in other issues, should we have at least the struct definition here?

File: exercises/practice/spiral-matrix/test_spiral_matrix.c (hunk 1 above)
Comment (wolf99) [reply exercises/practice/spiral-matrix/test_spiral_matrix.c:26]: How does this look when the tests run and have their standard output?
Would it interfere with how the web IDE detects pass/fail ?

File: exercises/practice/spiral-matrix/.meta/example.c [hunk 6]
```diff
@@ -0,0 +1,50 @@
+#include <stdlib.h>
+#include "spiral_matrix.h"
+
+spiral_matrix_t *spiral_matrix_create(int size)
+{
+   spiral_matrix_t *spiral_matrix = calloc(1, sizeof(spiral_matrix_t));
```
Comment (wolf99) [inline exercises/practice/spiral-matrix/.meta/example.c:6]: Small nit:
Consider refactoring type and variable names from `spiral_matrix` to `spiral` so that we then have a nice `spiral->matrix` during use?

**Review by wolf99 — COMMENTED** [review]
Sorry that it took me so long to review this!
Just a few small nits

File: exercises/practice/spiral-matrix/.meta/example.c (hunk 6 above)
Comment (ryanplusplus) [reply exercises/practice/spiral-matrix/.meta/example.c:6]: Sure, good idea

File: exercises/practice/spiral-matrix/.meta/example.c (hunk 3 above)
Comment (ryanplusplus) [reply exercises/practice/spiral-matrix/.meta/example.c:23]: I started with `size_t`, actually. The reason I changed to `int` is that I want to compare to `x + dx` and `y + dy` to `size` with a minimum of awkwardness and with `size_t` there's a signedness conflict. Casting `x + dx` to `size_t` isn't safe because `size_t` can't represent when `x + dx` is negative, casting `size` up to a larger signed type isn't an option, and casting `size` down to an `int` isn't much different than using `int` in the first place. I would use `ssize_t`, but this isn't included in standard C. C++ has standardized `std::ssize` and widespread use of unsigned size types is considered to have been a mistake in the C++ standard library.

In other words, I agree with you in principle, but this felt like the best of the bad options and least likely to cause awkward situations for students.

File: exercises/practice/spiral-matrix/spiral_matrix.h (hunk 5 above)
Comment (ryanplusplus) [reply exercises/practice/spiral-matrix/spiral_matrix.h:3]: Sure, I will add it.

File: exercises/practice/spiral-matrix/test_spiral_matrix.c (hunk 1 above)
Comment (ryanplusplus) [reply exercises/practice/spiral-matrix/test_spiral_matrix.c:26]: With this, the output for a failure is:

```
expected[0][0] = 1
actual[0][0] = 0
./test_spiral_matrix.c:52:test_trivial_spiral:FAIL: Expected TRUE Was FALSE
```

Without it, the output is:

```
./test_spiral_matrix.c:52:test_trivial_spiral:FAIL: Expected TRUE Was FALSE
```

>  Would it interfere with how the web IDE detects pass/fail ?

I don't think it will since the failure is still present, but I'm not sure whether it will show the console output in addition to the failure text.

Looking through the Unity docs I think I should be able to compare each row in the array. This should give enough context to understand where the failure is because each row is unique.

Comment (ryanplusplus) [reply exercises/practice/spiral-matrix/test_spiral_matrix.c]: With Unity's array matcher, the output is:

```
./test_spiral_matrix.c:22:test_spiral_of_size_5:FAIL: Element 2 Expected 25 Was 0
```

I think that's sufficient to work out where the error is even without knowing the row number.

**Comment by ryanplusplus** [comment]
> Sorry that it took me so long to review this! Just a few small nits

No worries, thanks for the review!

**Review by wolf99 — APPROVED** [review]
LGTM 👍
